File: CODE/DART/lib/processing.dart
// -- IMPORTS

import 'dart:convert';
import 'package:crypto/crypto.dart';

// -- FUNCTIONS

String getTextHash(
    String text
    )
{
    var byteArray = utf8.encode( text );
    var digest = md5.convert( byteArray );

    return digest.toString();
}

// ~~

String getTextUuid(
    String text
    )
{
    if ( text == '' )
    {
        return '';
    }
    else
    {
        var hash = getTextHash( text );

        return (
            hash.substring( 0, 8 )
            + '-'
            + hash.substring( 8, 12 )
            + '-'
            + hash.substring( 12, 16 )
            + '-'
            + hash.substring( 16, 20 )
            + '-'
            + hash.substring( 20, 32 )
            );
    }
}

// ~~

String getTextTuid(
    String text
    )
{
    if ( text == '' )
    {
        return '';
    }
    else
    {
        var tuid = base64Encode( md5.convert( utf8.encode( text ) ).bytes );

        return (
            tuid
                .replaceAll( '+', '-' )
                .replaceAll( '/', '_' )
                .replaceAll( '=', '' )
            );
    }
}

// ~~

String processPrimedText(
    String primedText
    )
{
    if ( primedText.startsWith( '‼#' ) )
    {
        return getTextUuid( primedText.substring( 2 ) );
    }
    else if ( primedText.startsWith( '‼%' ) )
    {
        return getTextTuid( primedText.substring( 2 ) );
    }
    else
    {
        return primedText;
    }
}

File: CODE/RUST/src/processing.rs
// -- IMPORTS

use base64::{engine::general_purpose::STANDARD, Engine as _};

// -- FUNCTIONS

pub fn get_text_hash(
    text: &str
    )
    -> String
{
    let byte_array = text.as_bytes();
    let digest = md5::compute( byte_array );

    hex::encode( digest.as_slice() )
}

// ~~

pub fn get_text_uuid(
    text: &str
    )
    -> String
{
    if text == ""
    {
        return "".to_string();
    }
    else
    {
        let hash = get_text_hash( text );

        return
            hash[ 0..8 ].to_string()
            + "-"
            + &hash[ 8..12 ]
            + "-"
            + &hash[ 12..16 ]
            + "-"
            + &hash[ 16..20 ]
            + "-"
            + &hash[ 20..32 ]
            ;
    }
}

// ~~

pub fn get_text_tuid(
    text: &str
    )
    -> String
{
    if text == ""
    {
        return "".to_string();
    }
    else
    {
        let digest = md5::compute( text.as_bytes() );
        let tuid = STANDARD.encode( digest.as_slice() );

        return
            tuid
                .replace( "+", "-" )
                .replace( "/", "_" )
                .replace( "=", "" )
            ;
    }
}

// ~~

pub fn process_primed_text(
    primed_text: &str
    )
    -> String
{
    if let Some( rest ) = primed_text.strip_prefix( "‼#" )
    {
        return format!( "\"{}\"", get_text_uuid( rest ) );
    }
    else if let Some( rest ) = primed_text.strip_prefix( "‼%" )
    {
        return format!( "\"{}\"", get_text_tuid( rest ) );
    }
    else
    {
        return format!( "\"{}\"", primed_text.replace( "\"", "\\\"" ) );
    }
}
